src/crimson/os/seastore/extent_reader.h

```diff
@@ -80,26 +89,30 @@ class ExtentReader {
     found_record_handler_t &handler    ///< [in] handler for records
   ); ///< @return used budget
 
-  void add_segment_manager(SegmentManager* segment_manager) {
-    assert(!segment_managers[segment_manager->get_device_id()] ||
-      segment_manager == segment_managers[segment_manager->get_device_id()]);
+  using release_ertr = SegmentManager::release_ertr;
+  release_ertr::future<> release_segment(segment_id_t id) {
+    assert(segment_managers[id.device_id()] != nullptr);
+    return segment_managers[id.device_id()]->release(id);
+  }
+
+  void add_segment_manager(SegmentManager* segment_manager, bool is_primary) {
+    ceph_assert(!segment_managers[segment_manager->get_device_id()]);
     segment_managers[segment_manager->get_device_id()] = segment_manager;
+    if (is_primary) {
+      ceph_assert(primary_segment_manager == nullptr);
+      primary_segment_manager = segment_manager;
+    }
   }
 
-  read_ertr::future<> read(
-    paddr_t addr,
-    size_t len,
-    ceph::bufferptr &out) {
-    assert(segment_managers[addr.get_device_id()]);
-    return segment_managers[addr.get_device_id()]->read(addr, len, out);
```

Comment:
From the standup, I think the consensus was that we need to split `read()` and `scanning()` interfaces into different classes, as `read()` would be common between `RBM` and `SegmentManager`, and scanning is currently only specific to `SegmentManager`.

IIUC this PR tries to achieve a similar goal with the `Device` class:
* `Device` is expected to be common between `RBM` and `SegmentManager`, so I think it can also contain the `read()` interface with other common interfaces, refer to https://github.com/ceph/ceph/blob/44b7db7006042b023a0b57c88e96bd5666b68d6a/src/crimson/os/seastore/device.h#L49-L112
* The read path will be changed to `Cache` -> `ExtentPlacementManager::read()` -(device-id)-> `Device::read()`, instead of `Cache` -> `ExtentReader::read()` -(device-id)-> `SegmentManager::read()`. I think it should be fine to extend `ExtentPlacementManager` to maintain both the placement decisions and the read paths as EPM manages and knows the placement information across all the devices.
* (WIP) I plan to rename the rest of `ExtentReader` class to `SegmentManagerGroup` to emphasize the fact that without the `read()` interface, it is responsible to manage a group of `SegmentManager`s for `SegmentCleaner` and `SegmentedJournal`. And `SegmentManagerGroup` will no longer be related to non-segment-based implementations (e.g. RBM device).
* Another consideration is that the current `release_segment()` is only implemented for the primary `SegmentManager`, not available to the other `SegmentManager`s. So I take the chance to extend the would-be `SegmentManagerGroup` class with `release_segment()` for all the `SegmentManager`s.

What do you think about it? @athanatos